### DISCUSSION TITLE
Defining sampling in problem-independent execution

### DISCUSSION THREAD

**Discussion by j-roh** [anyoptimization/pymoo#264] [Q&A] [answered]
How can I define the sampling method in problem-independent execution?

As far as I understood, by setting `sampling = get_sampling("int_random")` should work, but  I have encountered the problem that  float type Xs are included after the first generation. I was mimicking the problem-independent execution code as shown in https://pymoo.org/algorithms/usage.html?highlight=_evaluate and https://pymoo.org/operators/sampling.html?highlight=sampling to find the optimal solution, 

Below is the simplified code. 

```
import pandas as pd
import numpy as np
from pymoo.core.problem import Problem
from pymoo.core.evaluator import set_cv
from pymoo.algorithms.moo.nsga2 import NSGA2
from pymoo.factory import get_sampling



df_temp = pd.DataFrame(({"x1": [0, 1, 7, 98, -1],
                    "x2": [0, 8, -2, np.nan, np.nan],
                    "x3": [0, 4, 6, -3, np.nan],
                    "x4": [0, -4, np.nan, np.nan, np.nan],
                    "x5": [0, 7, 3, 21, -5]
                    }))

problem = Problem(n_var = 5, n_obj = 2, n_constr = 0,
                  xu = [4, 2, 3, 1, 4],
                  xl = np.zeros(5))

algorithm = NSGA2(pop_size = 3,
                  sampling = get_sampling("int_random"))

algorithm.setup(problem,
                verbose = True, 
                save_history = True)

for n_gen in range(10):
    pop = algorithm.ask()
    X = pop.get("X")
    print("X", X)
    f1, f2 = [], []
    for i in range(0, X.shape[0]):
        x_temp = X[i]
        xs = []
        for j in range(0, int(len(x_temp))):
            xs.append(df_temp.iloc[int(x_temp[j]), j]) # Just to show more than 2nd generation, I've added int() in this code
        f1.append(sum(xs))
        f2.append(xs[0] * xs[1] * xs[3] - xs[2])
    pop.set("F", np.column_stack([f1, f2]))
    set_cv(pop)
    algorithm.tell(infills = pop)

res = algorithm.result()

```

And from the second generation, float type Xs are included as shown in the [screenshot](https://imgur.com/a/u1hBVGM)

How can I edit the sampling method codes to get the Xs that are int type?

Thank you for your help in advance.

**Comment by blankjul** [comment, marked as answer]
Thanks for using pymoo!

`Sampling` is used ONLY to create the initial population. From there on, crossover and mutation are applied to create offsprings. 
We have designed the framework in a way that users have full flexibility for all operators. You can check out this tutorial for customizing crossover and mutation additionally: https://pymoo.org/problems/definition.html

A simple discrete optimization example is provided here: https://pymoo.org/customization/discrete.html

**Comment by j-roh** [comment]
Thank you so much for your reply and guide to the tutorials.

I was able to modify my codes by using them.

Another question I got is that , as far as I understood, by setting `type_var = int `should work change the variable types no matter what.

Below is the changed code where defined type_var under the class. And I could get all Xs in `int` as shown[ in the screenshot.](https://imgur.com/a/tql6G9k)
```
class MyProblem_2(Problem):
    
    def __init__(self):
        super().__init__(n_var=5, n_obj=2, n_constr=0, xl=np.zeros(5), xu=np.array([4, 2, 3, 1, 4]), type_var=int)

    def _evaluate(self, x, out, *args, **kwargs):
        global df_temp
        print(x)
        f1, f2 = [], []
        for i in range(0, x.shape[0]):
            x_temp = x[i]
            xs = []
            for j in range(0, int(len(x_temp))):
                xs.append(df_temp.iloc[int(x_temp[j]), j])
            f1.append(sum(xs))
            f2.append(xs[0] * xs[1] * xs[3] - xs[2])
        out["F"] = anp.column_stack([f1, f2])

```

The next one is the one I failed to get all Xs in  `int` , where I tried the similar way as I asked few days ago as shown [in the screenshot](https://imgur.com/a/JM8CGpk).

```
problem = Problem(n_var = 5, n_obj = 2, n_constr = 0,
                  xu = [4, 2, 3, 1, 4],
                  xl = np.zeros(5),
                  type_var = int)
```

I am not sure if I am understanding the definition method correctly and would be glad to know which I should study or if I am misunderstanding something.

Thank you so much in advance.

**Reply by blankjul** [reply]
The `type_var` is, at this point, only a type hint I have added for the test problem. This information will not be used during optimization.
You will have to use the appropriate parameters for your variable type. 
I am planning to have a better way of defining mixed variables in the next version, which will then have default settings for integer operators. At this point, the easiest way is to define the operators, as explained in my previous answer.

**Reply by j-roh** [reply]
Oh, that was the reason why it was explained as "(optional) A type hint." 

Thank you so much for your help. I really do appreciate it.
